vllm/v1/core/sched/scheduler.py

```diff
@@ -821,17 +822,35 @@ def update_from_output(
                 self._cached_reqs_data[req_data.req_id].append(req_data)
 
         self.running = new_running
-        engine_core_outputs = EngineCoreOutputs(
-            outputs=outputs,
-            scheduler_stats=self.make_stats(spec_decoding_stats),
-        )
-        if self.include_finished_set:
-            #TODO currently sending duplicates here, improve this
-            engine_core_outputs.finished_requests = (
-                scheduler_output.finished_req_ids | self.finished_req_ids)
+
+        engine_core_outputs = {
+            client_index: EngineCoreOutputs(outputs=outs)
+            for client_index, outs in outputs.items()
+        }
+
+        finished_req_ids = self.finished_req_ids_dict
+        if finished_req_ids is not None:
+            # Include ids of requests that finished since last outputs
+            # were sent.
+            for client_index, finished_set in finished_req_ids.items():
+                if (eco := engine_core_outputs.get(client_index)) is not None:
+                    eco.finished_requests = finished_set
+                else:
+                    engine_core_outputs[client_index] = EngineCoreOutputs(
+                        finished_requests=finished_set)
+            finished_req_ids.clear()
```

Comment:
Could you explain this change?

Reply:
Thanks @tlrmchlsmth 

This is adding each set of finished requests (arranged by client index) ids to the corresponding `EngineCoreOutputs`. If there isn't an `EngineCoreOutputs` for that client index (because there are no request outputs for that client index) then it adds an empty one containing just the `finished_requests` set.

I'll add some comments to make this clearer.